File: android/app/src/main/java/es/manabe/yomiyasu/core/services/DownloadManager.kt
package es.manabe.yomiyasu.core.services

import android.content.Context
import android.text.format.Formatter
import dagger.hilt.android.qualifiers.ApplicationContext
import es.manabe.yomiyasu.core.di.ApplicationScope
import es.manabe.yomiyasu.core.mokuro.MokuroParser
import es.manabe.yomiyasu.core.models.Book
import es.manabe.yomiyasu.core.models.Variant
import es.manabe.yomiyasu.core.networking.ApiClient
import es.manabe.yomiyasu.core.networking.Endpoint
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import java.io.File
import java.net.URLDecoder
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class DownloadRecord(
    val bookId: String,
    val visibleName: String,
    val variant: String,
    val downloadedAt: Long,
    val byteCount: Long,
    val pageCount: Int,
)

sealed interface DownloadState {
    data object NotDownloaded : DownloadState
    data object Queued : DownloadState
    data class Downloading(val progress: Double, val detail: String) : DownloadState
    data object Downloaded : DownloadState
    data class Failed(val message: String) : DownloadState

    val isActive: Boolean
        get() = this is Queued || this is Downloading
}

data class ActiveDownload(
    val bookId: String,
    val name: String,
    val progress: Double,
    val detail: String,
    val isQueued: Boolean,
)

@Singleton
class DownloadManager @Inject constructor(
    @ApplicationContext private val context: Context,
    private val api: ApiClient,
    @ApplicationScope private val scope: CoroutineScope,
) {
    private val rootDir: File = File(context.filesDir, "Yomiyasu/Downloads")

    private val _records = MutableStateFlow<Map<String, DownloadRecord>>(emptyMap())
    val records: StateFlow<Map<String, DownloadRecord>> = _records.asStateFlow()

    private val _states = MutableStateFlow<Map<String, DownloadState>>(emptyMap())
    val states: StateFlow<Map<String, DownloadState>> = _states.asStateFlow()

    private val mutex = Mutex()
    private val queue = ArrayDeque<Book>()
    private val bookNames = mutableMapOf<String, String>()
    private var activeBookId: String? = null
    private var currentJob: Job? = null

    init {
        rootDir.mkdirs()
        loadManifest()
    }

    val sortedRecords: List<DownloadRecord>
        get() = _records.value.values.sortedByDescending { it.downloadedAt }

    val totalBytes: Long
        get() = _records.value.values.sumOf { it.byteCount }

    val active: List<ActiveDownload>
        get() = _states.value.entries
            .filter { it.value.isActive }
            .mapNotNull { (bookId, state) ->
                val name = bookNames[bookId] ?: return@mapNotNull null
                when (state) {
                    is DownloadState.Queued -> ActiveDownload(bookId, name, 0.0, "En cola", true)
                    is DownloadState.Downloading ->
                        ActiveDownload(bookId, name, state.progress, state.detail, false)
                    else -> null
                }
            }

    fun stateFor(bookId: String): DownloadState =
        _states.value[bookId]
            ?: if (_records.value.containsKey(bookId)) DownloadState.Downloaded else DownloadState.NotDownloaded

    fun isDownloaded(bookId: String): Boolean = _records.value.containsKey(bookId)

    fun bookDirectory(bookId: String): File = File(rootDir, bookId)

    fun localHtmlFile(bookId: String): File = File(bookDirectory(bookId), "book.html")

    fun localEpubFile(bookId: String): File = File(bookDirectory(bookId), "book.epub")

    fun localImagesDirectory(bookId: String): File = File(bookDirectory(bookId), "images")

    fun enqueue(book: Book) {
        if (_records.value.containsKey(book.id)) return
        val currentState = _states.value[book.id]
        if (currentState?.isActive == true) return
        if (queue.any { it.id == book.id }) return

        updateState(book.id, DownloadState.Queued)
        bookNames[book.id] = book.visibleName
        queue.addLast(book)
        processQueue()
    }

    fun enqueueSeries(books: List<Book>) {
        books.forEach(::enqueue)
    }

    fun cancel(bookId: String) {
        queue.removeAll { it.id == bookId }

        if (activeBookId == bookId) {
            currentJob?.cancel()
        } else {
            updateState(bookId, DownloadState.NotDownloaded)
            bookNames.remove(bookId)
        }
    }

    fun delete(bookId: String) {
        cancel(bookId)
        bookDirectory(bookId).deleteRecursively()
        _records.value = _records.value - bookId
        updateState(bookId, DownloadState.NotDownloaded)
        saveManifest()
    }

    private fun processQueue() {
        if (currentJob != null || queue.isEmpty()) return

        val book = queue.removeFirst()
        activeBookId = book.id

        currentJob = scope.launch {
            try {
                performDownload(book)
            } finally {
                mutex.withLock {
                    currentJob = null
                    activeBookId = null
                    bookNames.remove(book.id)
                }
                processQueue()
            }
        }
    }

    private suspend fun performDownload(book: Book) {
        updateState(book.id, DownloadState.Downloading(0.0, "Preparando…"))

        try {
            when (book.variant) {
                Variant.Novela -> downloadNovel(book)
                else -> downloadManga(book)
            }
        } catch (cancellation: CancellationException) {
            updateState(book.id, DownloadState.NotDownloaded)
            bookDirectory(book.id).deleteRecursively()
            throw cancellation
        } catch (error: Exception) {
            updateState(book.id, DownloadState.Failed(error.message ?: "Error de descarga"))
            bookDirectory(book.id).deleteRecursively()
        }
    }

    private suspend fun downloadManga(book: Book) {
        val seriePath = book.seriePath ?: throw IllegalStateException("Libro sin ruta")
        val bookPath = book.path ?: throw IllegalStateException("Libro sin ruta")

        val htmlBytes = api.sendBytes(
            Endpoint.get("api/static/mangas/$seriePath/$bookPath.html"),
        )
        val html = htmlBytes.decodeToString()

        val parsed = withContext(Dispatchers.Default) { MokuroParser.parse(html) }

        val directory = bookDirectory(book.id)
        directory.mkdirs()
        localHtmlFile(book.id).writeBytes(htmlBytes)

        var bytes = htmlBytes.size.toLong()
        val pageCount = parsed.pages.size

        parsed.pages.forEachIndexed { index, page ->
            val decodedPath = runCatching { URLDecoder.decode(page.imagePath, "UTF-8") }
                .getOrDefault(page.imagePath)

            val data = api.sendBytes(Endpoint.get("api/static/mangas/$seriePath/$decodedPath"))

            val destination = File(localImagesDirectory(book.id), decodedPath)
            destination.parentFile?.mkdirs()
            destination.writeBytes(data)

            bytes += data.size

            val progress = (index + 1).toDouble() / maxOf(pageCount, 1).toDouble()
            val size = Formatter.formatShortFileSize(context, bytes)
            updateState(
                book.id,
                DownloadState.Downloading(
                    progress = progress,
                    detail = "${index + 1}/$pageCount pág. · $size",
                ),
            )
        }

        finish(
            DownloadRecord(
                bookId = book.id,
                visibleName = book.visibleName,
                variant = book.variant?.name?.lowercase() ?: "manga",
                downloadedAt = System.currentTimeMillis(),
                byteCount = bytes,
                pageCount = pageCount,
            ),
        )
    }

    private suspend fun downloadNovel(book: Book) {
        val seriePath = book.seriePath ?: throw IllegalStateException("Libro sin ruta")
        val bookPath = book.path ?: throw IllegalStateException("Libro sin ruta")

        val data = api.sendBytes(
            Endpoint.get("api/static/novelas/$seriePath/$bookPath.epub"),
        )

        val directory = bookDirectory(book.id)
        directory.mkdirs()
        localEpubFile(book.id).writeBytes(data)

        finish(
            DownloadRecord(
                bookId = book.id,
                visibleName = book.visibleName,
                variant = book.variant?.name?.lowercase() ?: "novela",
                downloadedAt = System.currentTimeMillis(),
                byteCount = data.size.toLong(),
                pageCount = 0,
            ),
        )
    }

    private fun finish(record: DownloadRecord) {
        _records.value = _records.value + (record.bookId to record)
        updateState(record.bookId, DownloadState.Downloaded)
        saveManifest()
    }

    private fun updateState(bookId: String, state: DownloadState) {
        _states.value = _states.value + (bookId to state)
    }

    private val manifestFile: File
        get() = File(rootDir, "manifest.json")

    private fun loadManifest() {
        try {
            if (!manifestFile.exists()) return
            val list = api.json.decodeFromString(
                ListSerializer(DownloadRecord.serializer()),
                manifestFile.readText(),
            )
            _records.value = list.associateBy { it.bookId }
        } catch (_: Exception) {
            // Manifest corrupto: se ignora y se reconstruirá con las siguientes descargas
        }
    }

    private fun saveManifest() {
        try {
            val data = api.json.encodeToString(
                ListSerializer(DownloadRecord.serializer()),
                sortedRecords,
            )
            manifestFile.writeText(data)
        } catch (_: Exception) {
            // Sin espacio o error de E/S: la descarga sigue siendo válida en disco
        }
    }
}

File: ios/Yomiyasu/Core/Services/DownloadManager.swift
import Foundation
import Observation
import os

struct DownloadRecord: Codable, Sendable, Identifiable, Equatable {
    let bookId: String
    let visibleName: String
    let variant: String
    let downloadedAt: Date
    let byteCount: Int64
    let pageCount: Int

    var id: String { bookId }
}

enum DownloadState: Equatable {
    case notDownloaded
    case queued
    case downloading(progress: Double, detail: String)
    case downloaded
    case failed(String)
}

struct ActiveDownload: Identifiable, Equatable {
    let bookId: String
    let name: String
    let progress: Double
    let detail: String
    let isQueued: Bool

    var id: String { bookId }
}

@MainActor
@Observable
final class DownloadManager {
    private(set) var records: [String: DownloadRecord] = [:]
    private(set) var states: [String: DownloadState] = [:]

    private let api: APIClient
    private let rootURL: URL
    private let logger = Logger(subsystem: "es.manabe.yomiyasu", category: "Downloads")

    private var queue: [Book] = []
    private var activeBookIds: [String] = []
    private var bookNames: [String: String] = [:]
    private var activeBookId: String?
    private var currentTask: Task<Void, Never>?

    init(api: APIClient, rootURL: URL? = nil) {
        self.api = api
        self.rootURL = rootURL ?? Self.defaultRootURL()

        try? FileManager.default.createDirectory(
            at: self.rootURL,
            withIntermediateDirectories: true
        )

        loadManifest()
    }

    private static func defaultRootURL() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory

        return base
            .appendingPathComponent("Yomiyasu", isDirectory: true)
            .appendingPathComponent("Downloads", isDirectory: true)
    }

    var sortedRecords: [DownloadRecord] {
        records.values.sorted { $0.downloadedAt > $1.downloadedAt }
    }

    var totalBytes: Int64 {
        records.values.reduce(0) { $0 + $1.byteCount }
    }

    var active: [ActiveDownload] {
        activeBookIds.compactMap { bookId in
            guard let name = bookNames[bookId] else { return nil }

            switch state(for: bookId) {
            case .queued:
                return ActiveDownload(
                    bookId: bookId,
                    name: name,
                    progress: 0,
                    detail: "En cola",
                    isQueued: true
                )
            case .downloading(let progress, let detail):
                return ActiveDownload(
                    bookId: bookId,
                    name: name,
                    progress: progress,
                    detail: detail,
                    isQueued: false
                )
            default:
                return nil
            }
        }
    }

    func state(for bookId: String) -> DownloadState {
        if let state = states[bookId] {
            return state
        }

        return records[bookId] != nil ? .downloaded : .notDownloaded
    }

    func isDownloaded(_ bookId: String) -> Bool {
        records[bookId] != nil
    }

    func bookDirectory(_ bookId: String) -> URL {
        rootURL.appendingPathComponent(bookId, isDirectory: true)
    }

    func localHTMLURL(for bookId: String) -> URL {
        bookDirectory(bookId).appendingPathComponent("book.html")
    }

    func localEpubURL(for bookId: String) -> URL {
        bookDirectory(bookId).appendingPathComponent("book.epub")
    }

    func localImagesDirectory(for bookId: String) -> URL {
        bookDirectory(bookId).appendingPathComponent("images", isDirectory: true)
    }

    func enqueue(_ book: Book) {
        guard records[book.id] == nil else { return }
        guard !queue.contains(where: { $0.id == book.id }) else { return }
        guard !(states[book.id]?.isActive ?? false) else { return }

        states[book.id] = .queued
        bookNames[book.id] = book.visibleName
        activeBookIds.append(book.id)
        queue.append(book)

        processQueue()
    }

    func enqueueSeries(_ books: [Book]) {
        for book in books {
            enqueue(book)
        }
    }

    func cancel(_ bookId: String) {
        queue.removeAll { $0.id == bookId }

        if activeBookId == bookId {
            currentTask?.cancel()
        } else {
            states[bookId] = .notDownloaded
            activeBookIds.removeAll { $0 == bookId }
            bookNames.removeValue(forKey: bookId)
        }
    }

    func delete(_ bookId: String) {
        cancel(bookId)
        try? FileManager.default.removeItem(at: bookDirectory(bookId))
        records.removeValue(forKey: bookId)
        states[bookId] = .notDownloaded
        saveManifest()
    }

    private func processQueue() {
        guard currentTask == nil, !queue.isEmpty else { return }

        let book = queue.removeFirst()
        activeBookId = book.id

        currentTask = Task { [weak self] in
            await self?.performDownload(book: book)
            self?.currentTask = nil
            self?.activeBookId = nil
            self?.activeBookIds.removeAll { $0 == book.id }
            self?.bookNames.removeValue(forKey: book.id)
            self?.processQueue()
        }
    }

    private func performDownload(book: Book) async {
        states[book.id] = .downloading(progress: 0, detail: "Preparando…")

        do {
            switch book.variant {
            case .novela:
                try await downloadNovel(book: book)
            default:
                if book.isImageFolder {
                    try await downloadImageFolder(book: book)
                } else {
                    try await downloadManga(book: book)
                }
            }
        } catch is CancellationError {
            logger.info("Descarga cancelada: \(book.visibleName, privacy: .public)")
            states[book.id] = .notDownloaded
            try? FileManager.default.removeItem(at: bookDirectory(book.id))
        } catch {
            logger.error("Descarga fallida: \(error.localizedDescription, privacy: .public)")
            states[book.id] = .failed(error.localizedDescription)
            try? FileManager.default.removeItem(at: bookDirectory(book.id))
        }
    }

    private func downloadManga(book: Book) async throws {
        guard let seriePath = book.seriePath, let bookPath = book.path else {
            throw APIError.unexpectedResponse
        }

        let htmlData = try await api.sendData(
            .get("api/static/mangas/\(seriePath)/\(bookPath).html")
        )

        guard let html = String(data: htmlData, encoding: .utf8) else {
            throw APIError.unexpectedResponse
        }

        let parsed = try await Task.detached(priority: .utility) {
            try MokuroParser.parse(html: html)
        }.value

        let directory = bookDirectory(book.id)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        try htmlData.write(to: localHTMLURL(for: book.id))

        var bytes = Int64(htmlData.count)

        for (index, page) in parsed.pages.enumerated() {
            try Task.checkCancellation()

            let decodedPath = page.imagePath.removingPercentEncoding ?? page.imagePath
            let data = try await api.sendData(
                .get("api/static/mangas/\(seriePath)/\(decodedPath)")
            )

            let destination = localImagesDirectory(for: book.id)
                .appendingPathComponent(decodedPath)
            try FileManager.default.createDirectory(
                at: destination.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            try data.write(to: destination)

            bytes += Int64(data.count)

            let progress = Double(index + 1) / Double(max(parsed.pages.count, 1))
            let size = ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file)
            states[book.id] = .downloading(
                progress: progress,
                detail: "\(index + 1)/\(parsed.pages.count) pág. · \(size)"
            )
        }

        finish(
            DownloadRecord(
                bookId: book.id,
                visibleName: book.visibleName,
                variant: book.variant?.rawValue ?? "manga",
                downloadedAt: .now,
                byteCount: bytes,
                pageCount: parsed.pages.count
            )
        )
    }

    /// Tomo sin mokuro: no hay html que parsear, el manifiesto de páginas
    /// viene en el detalle del libro (los listados no lo incluyen).
    private func downloadImageFolder(book: Book) async throws {
        guard let seriePath = book.seriePath else {
            throw APIError.unexpectedResponse
        }

        let detail: Book = try await api.send(.get("api/books/book/\(book.id)"))
        let pagePaths = detail.pagePaths ?? []

        guard !pagePaths.isEmpty else {
            throw APIError.unexpectedResponse
        }

        let directory = bookDirectory(book.id)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)

        var bytes: Int64 = 0

        for (index, fileName) in pagePaths.enumerated() {
            try Task.checkCancellation()

            let data = try await api.sendData(
                .get("api/static/mangas/\(seriePath)/\(fileName)")
            )

            // Misma estructura que en las descargas de mokuro: la imagen se
            // guarda bajo la carpeta del tomo, que es lo que espera el lector
            let destination = localImagesDirectory(for: book.id)
                .appendingPathComponent(
                    ImageFolderPages.joinedPath(
                        imagesFolder: detail.imagesFolder,
                        fileName: fileName
                    )
                )
            try FileManager.default.createDirectory(
                at: destination.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            try data.write(to: destination)

            bytes += Int64(data.count)

            let progress = Double(index + 1) / Double(max(pagePaths.count, 1))
            let size = ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file)
            states[book.id] = .downloading(
                progress: progress,
                detail: "\(index + 1)/\(pagePaths.count) pág. · \(size)"
            )
        }

        finish(
            DownloadRecord(
                bookId: book.id,
                visibleName: book.visibleName,
                variant: book.variant?.rawValue ?? "manga",
                downloadedAt: .now,
                byteCount: bytes,
                pageCount: pagePaths.count
            )
        )
    }

    private func downloadNovel(book: Book) async throws {
        guard let seriePath = book.seriePath, let bookPath = book.path else {
            throw APIError.unexpectedResponse
        }

        let data = try await api.sendData(
            .get("api/static/novelas/\(seriePath)/\(bookPath).epub")
        )

        try Task.checkCancellation()

        let directory = bookDirectory(book.id)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        try data.write(to: localEpubURL(for: book.id))

        finish(
            DownloadRecord(
                bookId: book.id,
                visibleName: book.visibleName,
                variant: book.variant?.rawValue ?? "novela",
                downloadedAt: .now,
                byteCount: Int64(data.count),
                pageCount: 0
            )
        )
    }

    private func finish(_ record: DownloadRecord) {
        records[record.bookId] = record
        states[record.bookId] = .downloaded
        saveManifest()
        logger.info("Descarga completada: \(record.visibleName, privacy: .public)")
    }

    private var manifestURL: URL {
        rootURL.appendingPathComponent("manifest.json")
    }

    private func loadManifest() {
        guard let data = try? Data(contentsOf: manifestURL),
              let list = try? JSONDecoder().decode([DownloadRecord].self, from: data) else {
            return
        }

        records = Dictionary(uniqueKeysWithValues: list.map { ($0.bookId, $0) })
    }

    private func saveManifest() {
        let list = sortedRecords

        guard let data = try? JSONEncoder().encode(list) else { return }
        try? data.write(to: manifestURL)
    }
}

private extension DownloadState {
    var isActive: Bool {
        switch self {
        case .queued, .downloading: true
        default: false
        }
    }
}
